Optimizing contacts activity API (refactoring of PR-10237 for Mautic v5)

Moongazer opened the pull request:
This is a refactoring of [PR-10237](https://github.com/mautic/mautic/pull/10237) for Mautic v5 as [requested](https://github.com/mautic/mautic/pull/12187#issuecomment-1504948090). See original PR for feature and testing description.

Info: This refactoring covers 100% of the original PR, except 1 file which have been skipped because it exists already in 5.x branch: `app/bundles/CoreBundle/Doctrine/PreUpAssertionMigration.php`.

Fixes https://github.com/mautic/mautic/issues/12146

## Description

Add indicies to columns ( especially those that use file sort) plus other fixes:

- Set the event type name to default in the beginning of each for loop (PageBundle/EventListener/LeadSubscriber::onTimelineGenerate)
- Use second parameter for ordering in TimelineTrait
- Remove 'where stat.lead_id = null' from EmailBundle/Entity/EmailReplyRepository::getByLeadForTimeline
- USE INDEX in CampaignBundle/Entity/LeadEventLogRepository::getLeadLogs

## Steps to test this PR

1. Load up [this PR](https://gitpod.io/#https://github.com/mautic/mautic/pull/12305)
2. Test that the API for all affected parts work as before (https://developer.mautic.org/#rest-api), mainly:

/api/contacts/2/activity
/api/contacts/activity

3. Check the migrations work (if possible!)

Moongazer committed:
Optimizing contacts activity API (refactoring of PR-10237 for Mautic v5)
Manually resolved conflicts during rebasing fix/issue-12146 with current 5.x branch

escopecz commented:
@Moongazer are you able to resolve the conflict?

Moongazer commented:
@escopecz Jap, conflict resolved

RCheesley reviewed (dismissed):
Tested that the API works in both cases mentioned with this PR applied.

Will need code review, and attention to migrations, possibly?

RCheesley commented:
Hi @Moongazer could you please rebase this to 5.1 so we can get it merged with that release? Thanks!

AvinashDalvi89 commented:
Tested APIs looks fine.

AvinashDalvi89 reviewed (approved):
Tested two mentioned APIs endpoint and its working fine as expected.

Moongazer commented:
> Hi @Moongazer could you please rebase this to 5.1 so we can get it merged with that release? Thanks!

By accident I rebased it to the current 5.x branch. Can I perform another rebase from here to 5.1, or will it destroy something? (@escopecz)

Moongazer commented:
PS: Just by chance I saw the following line where I'm wondering:

[\Mautic\PageBundle\EventListener\LeadSubscriber::onTimelineGenerate() line 66](https://github.com/Moongazer/mautic/blob/8970669556defe7d085fb4a02b8bcd263b4c108f/app/bundles/PageBundle/EventListener/LeadSubscriber.php#L66):
`$event->addSerializerGroup('pageList', 'hitDetails');`

Looking at the file [\Mautic\LeadBundle\Event\LeadTimelineEvent::addSerializerGroup($group)](https://github.com/Moongazer/mautic/blob/8970669556defe7d085fb4a02b8bcd263b4c108f/app/bundles/LeadBundle/Event/LeadTimelineEvent.php#L519) line 519, we see that there is only ONE parameter `$group`, which can be an array. IMO that means, that the call should pass `['pageList', 'hitDetails']` as value, or am I wrong here? Did this method changed or how was it working during the tests?

escopecz commented:
@Moongazer yes, use 5.1 as target and base branch. You can edit the target branch when you edit the title of this PR.

escopecz commented:
And I understand the code the same as you do. It should be an array if there is more than 1 param.

Moongazer committed:
Merge branch 'mautic:5.x' into fix/issue-12146

Moongazer committed:
Fix do not pass multiple arguments to LeadTimelineEvent::addSerializerGroup() but using an array instead

Moongazer commented:
> @Moongazer yes, use 5.1 as target and base branch. You can edit the target branch when you edit the title of this PR.

All right, I changed it to 5.1 above. Also I added another commit to pass the parameters for `LeadTimelineEvent::addSerializerGroup()` as array.

escopecz commented:
Nice! There seem to be some conflicts. But perhaps wait because we'll have to merge 5.x into 5.1 which should resolve them.

escopecz committed:
Merge branch '5.x' into fix/issue-12146

RCheesley commented:
@Moongazer could you resolve the conflicts please? Thanks!

Moongazer committed:
Merge branch '5.x' into fix/issue-12146

RCheesley committed:
Merge branch '5.x' into fix/issue-12146

RCheesley commented:
Looks like there are some test failures to address with this PR folks!

escopecz commented on app/bundles/LeadBundle/Entity/StagesChangeLog.php:
```diff
@@ -44,6 +44,7 @@ public static function loadMetadata(ORM\ClassMetadata $metadata): void
 
         $builder->setTable('lead_stages_change_log')
             ->setCustomRepositoryClass(StagesChangeLogRepository::class);
+            ->addIndex(['date_added'], 'lead_stages_change_log_date_added');
```
`PHP Parse error:  syntax error, unexpected token "->" in /home/runner/work/mautic/mautic/app/bundles/LeadBundle/Entity/StagesChangeLog.php on line 47`
Remove the semicolon on the previous line.

Moongazer replied:
Done

Moongazer committed:
Update StagesChangeLog.php

Fix wrong colon

patrykgruszka reviewed (changes requested):
It is worth noting that these changes also affect contacts activity in the admin panel, not just the API. I've tested the "up" migration procedure and it works fine. I did some smoke tests in the panel with the timelines and had no problems. 

I haven't been able to test the "down" migration as it is not implemented. I think this should be included to ensure that the migration is fully reversible.

RCheesley commented:
@Moongazer if you agree with the suggestion, could you add the 'down' migration per @patrykgruszka 's feedback?

Moongazer commented:
@RCheesley @All

Actually I've no idea what the up/down thing is about, because I was just refactoring someone else's pull-request (https://github.com/mautic/mautic/pull/10237), without looking into what it does. Also - very unfortunately - I'm overwhelmed busy atm with no end in sight, so I'd ask if someone else can take over this task (or, if it's a simple fix, just posting a ready code snipped here which I could push than). Thanks for our understanding.

escopecz commented on app/migrations/Version20230506112314.php:
```diff
@@ -0,0 +1,33 @@
+        }, sprintf('The index "%s" has already been added to the table "%s".', $this->getIndexName(), $this->getTableName()));
+    }
+
+    public function up(Schema $schema): void
+    {
+        $this->addSql(sprintf('CREATE INDEX %s ON %s (date_added)', $this->getIndexName(), $this->getTableName()));
+    }
```
The down() method is undoing what the up() method does in case someone needs to downgrade. ChatGPT can help you write it if you provide the up() method. Example:
```suggestion
    public function up(Schema $schema): void
    {
        $this->addSql(sprintf('CREATE INDEX %s ON %s (date_added)', $this->getIndexName(), $this->getTableName()));
    }
    
    public function down(Schema $schema): void
    {
        $this->addSql("DROP INDEX {$this->getIndexName()} ON {$this->getTableName()}");
    }
```

RCheesley reviewed (dismissed):
API working as expected with the latest changes.

RCheesley committed:
Merge branch '5.x' into fix/issue-12146

Moongazer committed:
Update app/migrations/Version20230506112314.php

Co-authored-by: John Linhart <jan@linhart.email>

Moongazer committed:
Merge branch 'mautic:5.x' into fix/issue-12146

RCheesley committed:
Merge branch '5.x' into fix/issue-12146

RCheesley commented:
@patrykgruszka @escopecz could you please re-review this one now the migration is in place? Would be super to get this merged! :)

escopecz commented:
@Moongazer could you please resolve the Code Style issue?

escopecz reviewed (dismissed):
I see no problem in the code. I think the down() method in this case is not necessary as the indexes won't hurt if they stay even during the Mautic downgrade.

I've run the migrations and they all run successfully. I've executed the [API Library tests](https://github.com/mautic/api-library/actions/runs/9959726646) on this PR and it passed as well.

Only the CS issue must be fixed.

Moongazer committed:
Merge branch 'mautic:5.x' into fix/issue-12146

Moongazer committed:
Fix CS-Fixer error in app/migrations/Version20230506112314.php

Moongazer commented:
@escopecz CS error is fixed, ready to merge

escopecz reviewed (approved):
We'll need one more approval to get this merged.

patrykgruszka reviewed (approved):
I see no issues in the code, and it works fine.

escopecz committed:
Merge branch '5.x' into fix/issue-12146
